### PR TITLE
Transaction Sharing across Connectors

### PR DESCRIPTION
**Pull request by sachdevs**
Do not review, just for tracking purposes.

See referenced PR for most of the details.

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/PrismTransactionHandle.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.facebook.presto.hive;
+
+public class PrismTransactionHandle
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/PrismTransactionHandle.java]: We can't expose prism concept to open source. Instead, override `HiveTransactionManager` internally.

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java [hunk 2]
```diff
@@ -417,6 +418,14 @@ public void checkOpenTransaction()
                 if (catalog.isPresent()) {
                     registerCatalog(catalog.get());
                 }
+
+                if (catalogManager.hasFederations(catalogName)) {
+                    Set<Optional<Catalog>> federatedCatalogs = catalogManager.getFederatedCatalogs(catalogName);
+                    if (!federatedCatalogs.stream().allMatch(Optional::isPresent)) {
+                        throw new IllegalStateException("Not all federated catalogs exist");
+                    }
+                    federatedCatalogs.forEach(federatedCatalog -> registerCatalog(federatedCatalog.get()));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java]: This isn't necessary. `CatalogManager` is injected. All catalogs should be visible as long as they are installed.

Comment (sachdevs) [reply presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java]: So the purpose of this is to expose a set of "federated" catalogs. Catalogs are indeed visible but the visibility is not the issue - the issue is that we need a way to know which set of catalogs should be installed for a given connector. I.e. when a hive transaction starts, it does not care for scuba transactions at all. Without this we would have to arbitrarily allow access to any catalogs in catalog manager which may be unsafe due to any ACL logic.

Might be a better way to do this though.

File: presto-hive/src/main/java/com/facebook/presto/hive/PrismTransactionHandle.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+        extends HiveTransactionHandle
+{
+    private final TransactionalMetadata transactionalMetadata;
```
Comment (jessesleeping) [inline presto-hive/src/main/java/com/facebook/presto/hive/PrismTransactionHandle.java]: In general `ConnectorTransactionHandle` should be lightweight and JSON serializable. Including the entire `TransactionalMetadata` object as a handle seems not right.

Comment (sachdevs) [reply presto-hive/src/main/java/com/facebook/presto/hive/PrismTransactionHandle.java]: Yeah so I just made the TransactionalMetadata lie under a package-private method now (see presto-fb PR) and Jackson cant access it (We shouldnt be giving public access to this hack anyway). So it will serialize but it'll fail if we try to send transactionhandle over the wire for hive which we dont do.

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java [hunk 4]
```diff
@@ -90,8 +90,9 @@
     private final Executor finishingExecutor;
 
     private final Map<String, FunctionNamespaceManager<?>> functionNamespaceManagers = new HashMap<>();
+    private final ImmutableMap<String, String> extraCatalogs;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java]: nit: `Map<String, String>`. Let's call it `companionCatalogs`. Add a comment to explain that "Some catalog's transaction will be delegated to catalog. Use this enforce both transactions are present".

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java [hunk 5]
```diff
@@ -344,6 +346,8 @@ public void fail(TransactionId transactionId)
         private final Map<String, FunctionNamespaceManager<?>> functionNamespaceManagers;
         @GuardedBy("this")
         private final Map<String, FunctionNamespaceTransactionMetadata> functionNamespaceTransactions = new ConcurrentHashMap<>();
+        // Need not be guarded since immutable
+        private final ImmutableMap<String, String> extraCatalogs;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java]: same `Map`

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java [hunk 6]
```diff
@@ -344,6 +346,8 @@ public void fail(TransactionId transactionId)
         private final Map<String, FunctionNamespaceManager<?>> functionNamespaceManagers;
         @GuardedBy("this")
         private final Map<String, FunctionNamespaceTransactionMetadata> functionNamespaceTransactions = new ConcurrentHashMap<>();
+        // Need not be guarded since immutable
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java]: nah, no need to have the comment

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java [hunk 7]
```diff
@@ -352,7 +356,8 @@ public TransactionMetadata(
                 boolean autoCommitContext,
                 CatalogManager catalogManager,
                 Executor finishingExecutor,
-                Map<String, FunctionNamespaceManager<?>> functionNamespaceManagers)
+                Map<String, FunctionNamespaceManager<?>> functionNamespaceManagers,
+                ImmutableMap<String, String> extraCatalogs)
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java]: Map

File: presto-tests/src/main/java/com/facebook/presto/tests/H2QueryRunner.java [hunk 8]
```diff
@@ -200,6 +200,11 @@ public MaterializedResult execute(Session session, @Language("SQL") String sql,
         return materializedRows;
     }
 
+    public Handle getHandle()
```
Comment (highker) [inline presto-tests/src/main/java/com/facebook/presto/tests/H2QueryRunner.java]: Separate this to another commit

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java [hunk 9]
```diff
@@ -417,6 +423,12 @@ public void checkOpenTransaction()
                 if (catalog.isPresent()) {
                     registerCatalog(catalog.get());
                 }
+
+                if (extraCatalogs.containsKey(catalogName)) {
+                    Optional<Catalog> federatedCatalog = catalogManager.getCatalog(extraCatalogs.get(catalogName));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java]: companionCatalog

File: presto-main/src/main/java/com/facebook/presto/transaction/TransactionManagerConfig.java [hunk 10]
```diff
@@ -72,4 +80,18 @@ public TransactionManagerConfig setMaxFinishingConcurrency(int maxFinishingConcu
...
         return this;
     }
+
+    @NotNull
+    public Map<String, String> getExtraCatalogs()
+    {
+        return this.accessibleCatalogs;
+    }
+
+    @Config("transaction.extra-accessible-catalogs")
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/transaction/TransactionManagerConfig.java]: same

**Comment by sachdevs** [comment]
Addressed comments.
